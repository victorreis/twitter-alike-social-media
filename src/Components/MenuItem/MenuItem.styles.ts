import { NavLink } from 'react-router-dom';

import styled from 'styled-components';

import { MenuItemStyleProps } from './MenuItem.types';

export const MenuItemContainer = styled(NavLink)<MenuItemStyleProps>`
  color: ${({ theme }) => theme.colors.font.default};
  ${({ theme }) => theme.typographies.h4};
  display: flex;
  text-decoration: none;
  column-gap: 1rem;
`;
