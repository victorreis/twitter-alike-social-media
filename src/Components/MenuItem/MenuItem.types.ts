import { TestProps } from '../../Config/Tests/Test.types';

export interface RequiredMenuItemProps {
  /**
   * Component's children.
   */
  children: React.ReactNode;

  /**
   * Path that will be used to redirect the user to another page.
   */
  to: string;
}

export interface DefaultMenuItemProps {}

export interface OptionalMenuItemProps {}

export type MenuItemProps = RequiredMenuItemProps &
  DefaultMenuItemProps &
  OptionalMenuItemProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLAnchorElement>, 'children'>;

export type MenuItemStyleProps = Required<DefaultMenuItemProps>;
