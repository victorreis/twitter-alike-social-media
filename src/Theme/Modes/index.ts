export { darkTheme } from './Dark.theme';
