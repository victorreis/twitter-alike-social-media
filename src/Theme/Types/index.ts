export * from './Borders.types';
export * from './Colors.types';
export * from './CustomTheme.types';
export * from './Sizes.types';
export * from './Typographies.types';
