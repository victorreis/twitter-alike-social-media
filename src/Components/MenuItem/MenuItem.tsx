import { TestProps } from '../../Config/Tests/Test.types';
import { MenuItemContainer } from './MenuItem.styles';
import { MenuItemProps, DefaultMenuItemProps } from './MenuItem.types';

export const menuItemDefaults: Required<DefaultMenuItemProps> &
  Required<TestProps> = {
  testID: 'MenuItem',
};

export const MenuItem: React.FC<MenuItemProps> = (props): JSX.Element => {
  const {
    testID = menuItemDefaults.testID,
    children,
    to,
    style,
    ...others
  } = props;

  return (
    <MenuItemContainer data-testid={testID} style={style} to={to} {...others}>
      {children}
    </MenuItemContainer>
  );
};
