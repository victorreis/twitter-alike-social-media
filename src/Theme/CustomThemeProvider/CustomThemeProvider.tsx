/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
