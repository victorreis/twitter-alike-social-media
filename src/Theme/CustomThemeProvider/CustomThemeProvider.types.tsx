import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
