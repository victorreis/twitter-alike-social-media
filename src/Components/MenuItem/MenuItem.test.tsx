import {
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { hexToRgb } from '../../Utils/Transform/hexToRgb.util';
import { MenuItem, menuItemDefaults } from './MenuItem';
import { RequiredMenuItemProps, MenuItemProps } from './MenuItem.types';

describe('menuItem component tests', () => {
  const children = 'text';
  const to = '/user/teste';

  const requiredProps: RequiredMenuItemProps = {
    children,
    to,
  };

  const setup = (props?: MenuItemProps) => {
    const renderRTR = () =>
      renderRTRCreator<MenuItemProps>(MenuItem, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<MenuItemProps>(MenuItem, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(menuItemDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the text`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const element = screen.getByText(children);

      expect(element).toBeInTheDocument();
    });
  });

  describe('style tests', () => {
    it(`should have styled the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(menuItemDefaults.testID);

      expect(container).toHaveStyle({
        color: hexToRgb(themes.default.colors.font.default),
        ...themes.default.typographies.h4,
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
