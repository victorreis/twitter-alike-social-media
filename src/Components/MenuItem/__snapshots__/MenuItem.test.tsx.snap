// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menuItem component tests snapshot tests should render correctly 1`] = `
<a
  className="MenuItemContainer-twitter-alike-social-media__sc-87xn0-0 hUZhVl"
  data-testid="MenuItem"
  href="/user/teste"
  onClick={[Function]}
>
  text
</a>
`;
